atomicapp/nulecule/main.py

```diff
@@ -0,0 +1,234 @@
+# -*- coding: utf-8 -*-
+import anymarkup
+import copy
+import logging
+import os
+import shutil
+import uuid
+from string import Template
+
+from atomicapp.constants import (APP_ENT_PATH, EXTERNAL_APP_DIR,
+                                 GLOBAL_CONF, CACHE_DIR)
+from atomicapp.nulecule.lib import NuleculeBase
+from atomicapp.nulecule.container import DockerHandler
+from atomicapp.utils import Utils
+
+logger = logging.getLogger(__name__)
+
+
+class NuleculeManager(object):
+
+    def __init__(self, image, unpack_path=None):
+        self.image = image
+        self.initialize(unpack_path)
+
+    def initialize(self, unpack_path=None):
+        if unpack_path:
+            self.unpack_path = unpack_path
+        else:
+            self.app_name = '{}-{}'.format(
+                Utils.sanitizeName(self.image), uuid.uuid1())
+            self.nulecule = None
+            self.unpack_path = os.path.join(CACHE_DIR, self.app_name)
+        self.nulecule = None
+
+    def unpack(self, update=False):
+        if not os.path.isdir(self.unpack_path) or update:
+            self.nulecule = Nulecule.unpack(self.image, self.unpack_path)
+
+    def install(self):
+        self.unpack()
+        self.nulecule.install()
+
+    def run(self, provider_key=None, dry=False):
+        self.install()
+        self.nulecule.run(provider_key, dry)
+
+    def stop(self, provider_key=None):
+        self.nulecule.stop(provider_key)
+
+    def uninstall(self):
+        self.stop()
+        self.nulecule.uninstall()
+
+    def clean(self, force=False):
+        self.uninstall()
+        shutil.rmtree(self.unpack_path)
+        self.initialize()
+
+
+class Nulecule(NuleculeBase):
+    """
+    This represents an application compliant with Nulecule specification.
+    A Nulecule instance can have instances of Nulecule and Nulecule as
+    components. A Nulecule instance knows everything about itself and its
+    componenents, but does not have access to it's parent's scope.
+    """
+    def __init__(self, id, specversion, metadata, graph, basepath,
+                 requirements=None, params=None, config=None,
+                 namespace=GLOBAL_CONF):
+        self.id = id
+        self.specversion = specversion
+        self.metadata = metadata
+        self.graph = graph
+        self.basepath = basepath
+        self.requirements = requirements
+        self.params = params or []
+        self.namespace = namespace
+        self.config = None
+
+    @classmethod
+    def unpack(cls, image, path, config={}, namespace=GLOBAL_CONF):
+        docker_handler = DockerHandler()
+        docker_handler.pull(image)
+        docker_handler.extract(image, APP_ENT_PATH, path)
+        nulecule_data = anymarkup.parse_file(
+            os.path.join(path, 'Nulecule'))
+        nulecule = Nulecule(config=config, basepath=path, namespace=namespace,
+                            **nulecule_data)
+        nulecule.load_components()
+        return nulecule
```

Comment:
At some point we need to make it extremely obvious to the user what directory we are unpacking to. A print statement here would be a first start.. we might also want to consider printing out a very human readable message right before we return that lets them know as well.